File: functionTests/tfunHDDCTests/paperVerificationTest.R
library(TFunHDDC)
data = fitNOxBenchmark(15)
labels = fitNOxBenchmark()$groupd

K = 2
inits = c('random', 'vector', 'mini-em', 'kmeans')
vec = rep(2, length(labels))
vec[1:length(labels)/2] = 1
itTest = 1

cl=list()
for(i in 1:K) cl[[as.character(i)]] = rep(0, length(labels))

statsrand = list(bic = c(), time=0, cls = cl, conv = 0, ccr = c(), ccrStdev=0, ari=c(), ariStdev=0)
statsvec = list(bic = c(), time=0, cls = cl, conv = 0, ccr = c(), ccrStdev=0, ari=c(), ariStdev=0)
statsmini = list(bic = c(), time=0, cls = cl, conv = 0, ccr = c(), ccrStdev=0, ari=c(), ariStdev=0)
statsk = list(bic = c(), time=0, cls = cl, conv = 0, ccr = c(), ccrStdev=0, ari=c(), ariStdev=0)

j = 0
myDefault = list()
myDefault$iter.max = 10
myDefault$nstart = 1
myDefault$algorithm = c("Hartigan-Wong", "Lloyd", "Forgy","MacQueen")
myDefault$trace = FALSE
myDefault$alpha = 0.2
#kmc = list(iter.max=10, nstart=1, algorithm=c("Hartigan-Wong"))

start = Sys.time()

res = TFunHDDC::tfunHDDC(data$fd, K=2, model=c("AKJBKQKDK", "AKBKQKDK", "AKBQKDK", "AKJBQKDK", "ABKQKDK", "ABQKDK"), init = 'kmeans', threshold = 0.1, nb.rep = 20, dfconstr = 'no')

print(Sys.time() - start)

# for(i in inits){
#   print(i)
#   j = 1
#   
#   res = TFunHDDC::tfunHDDC(data, K=K, model=c("AKJBKQKDK", "AKBKQKDK", "AKBQKDK", "AKJBQKDK", "ABKQKDK", "ABQKDK"), init=i, threshold=0.6, nb.rep=20, init.vector=vec)
#   if (!inherits(res, 't-funHDDC')){
#     next
#   }
#   
#   print(paste0("CCR for init=", i))
#   print(sum(diag(table(res$class, labels)))/length(labels))
# }
print(table(labels, res$class))
print(TFunHDDC:::.T_hddc_ari(labels, res$class))

File: functionTests/tfunHDDCTests/paperVerificationTest.py
import sys
sys.path.append('../..')
import time
import numpy as np
import tfunHDDC as tfun
import NOxBenchmark as NOx
import matplotlib.pyplot as plt
import csv
import sklearn
from copy import *

if __name__ == '__main__':
    K = 2
    coreNum = 16
    inits=['kmeans']
    data = NOx.fitNOxBenchmark()['data']
    labels = NOx.fitNOxBenchmark()['target'].astype(int)
    vec = np.repeat(1, len(data.coefficients))
    vec[0:int(len(data.coefficients)/2)] = 0
    kmeans_control1 = {'max_iter': 20, 'n_init': 1, 'algorithm': 'lloyd'}
    kmeans_control2 = {'max_iter':20, 'n_init': 1, 'algorithm': 'elkan'}
    kmeans_control = [kmeans_control1, kmeans_control2]
    itTest = 1

    cl = {}
    for i in range(K):
        cl[f'{i}'] = np.zeros(len(data.coefficients))

    statsrand = {'time': 0, 'cl': deepcopy(cl), 'ccr':np.zeros(itTest), 'ccrStdev':0, 'ari': np.zeros(itTest), 'ariStdev':0}
    statsvec = {'time': 0, 'cl': deepcopy(cl), 'ccr':np.zeros(itTest), 'ccrStdev':0, 'ari':np.zeros(itTest), 'ariStdev':0}
    statsmini = {'time': 0, 'cl': deepcopy(cl), 'ccr':np.zeros(itTest), 'ccrStdev':0, 'ari':np.zeros(itTest), 'ariStdev':0}
    statsk = {'time': 0, 'cl': deepcopy(cl), 'ccr':np.zeros(itTest), 'ccrStdev':0, 'ari':np.zeros(itTest), 'ariStdev':0}
    
    for i in inits:
        
        if i == 'kmeans':
            for k in kmeans_control:
                res = tfun.tfunHDDC(data, model=["AKJBKQKDK", "AKBKQKDK", "AKBQKDK", "AKJBQKDK", "ABKQKDK", "ABQKDK"], K=K, mc_cores=coreNum, threshold=0.6, nb_rep = 20, init=i, init_vector=vec, kmeans_control=k)
                if not isinstance(res, tfun.TFunHDDC):
                    print(res)
                    continue
                print(f"CCR for init={i}, 'algo:{k['algorithm']}")
                print(np.sum(np.diag(sklearn.metrics.confusion_matrix(labels, res.cl, labels=[k for k in range(K)])))/len(data.coefficients))
            
        else:
            res = tfun.tfunHDDC(data, model=["AKJBKQKDK", "AKBKQKDK", "AKBQKDK", "AKJBQKDK", "ABKQKDK", "ABQKDK"], K=K, mc_cores=coreNum, threshold=0.6, nb_rep = 20, init=i, init_vector=vec)
            if not isinstance(res, tfun.TFunHDDC):
                print(res)
                continue
            print(f"CCR for init={i}")
            print(np.sum(np.diag(sklearn.metrics.confusion_matrix(labels, res.cl, labels=[k for k in range(K)])))/len(data.coefficients))